File: moqt/internal/message/subscribe.go
package message

import (
	"io"
)

/*
* SUBSCRIBE Message {
*   Message Length (varint)
*   Subscribe ID (varint)
*   Broadcast Path (string)
*   Track Name (string)
*   Subscriber Priority (varint)
*   Subscriber Ordered (varint)
*   Subscriber Max Latency (varint)
*   Start Group (varint)
*   End Group (varint)
* }
*
* Broadcast Path and Track Name are length-prefixed UTF-8 strings.
* Start Group and End Group use 0 for the default/latest and unbounded values.
 */
type SubscribeMessage struct {
	SubscribeID          uint64
	BroadcastPath        string
	TrackName            string
	SubscriberPriority   uint8
	SubscriberOrdered    uint8
	SubscriberMaxLatency uint64
	StartGroup           uint64
	EndGroup             uint64
}

func (s SubscribeMessage) Len() int {
	var l int

	l += VarintLen(uint64(s.SubscribeID))
	l += StringLen(s.BroadcastPath)
	l += StringLen(s.TrackName)
	l += VarintLen(uint64(s.SubscriberPriority))
	l += VarintLen(uint64(s.SubscriberOrdered))
	l += VarintLen(s.SubscriberMaxLatency)
	l += VarintLen(s.StartGroup)
	l += VarintLen(s.EndGroup)

	return l
}

func (s SubscribeMessage) Encode(w io.Writer) error {
	msgLen := s.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, uint64(s.SubscribeID))
	b, _ = WriteVarint(b, uint64(len(s.BroadcastPath)))
	b = append(b, s.BroadcastPath...)
	b, _ = WriteVarint(b, uint64(len(s.TrackName)))
	b = append(b, s.TrackName...)
	b, _ = WriteVarint(b, uint64(s.SubscriberPriority))
	b, _ = WriteVarint(b, uint64(s.SubscriberOrdered))
	b, _ = WriteVarint(b, s.SubscriberMaxLatency)
	b, _ = WriteVarint(b, s.StartGroup)
	b, _ = WriteVarint(b, s.EndGroup)

	_, err := w.Write(b)
	return err
}

func (s *SubscribeMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	s.SubscribeID = num
	b = b[n:]

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	s.BroadcastPath = str
	b = b[n:]

	str, n, err = ReadString(b)
	if err != nil {
		return err
	}
	s.TrackName = str
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	s.SubscriberPriority = uint8(num)
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	s.SubscriberOrdered = uint8(num)
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	s.SubscriberMaxLatency = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	s.StartGroup = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	s.EndGroup = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/subscribe.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface SubscribeMessageInit {
	subscribeId?: number;
	broadcastPath?: string;
	trackName?: string;
	subscriberPriority?: number;
	subscriberOrdered?: number;
	subscriberMaxLatency?: number;
	startGroup?: number;
	endGroup?: number;
}

export class SubscribeMessage {
	subscribeId: number;
	broadcastPath: string;
	trackName: string;
	subscriberPriority: number;
	subscriberOrdered: number;
	subscriberMaxLatency: number;
	startGroup: number;
	endGroup: number;

	constructor(init: SubscribeMessageInit = {}) {
		this.subscribeId = init.subscribeId ?? 0;
		this.broadcastPath = init.broadcastPath ?? "";
		this.trackName = init.trackName ?? "";
		this.subscriberPriority = init.subscriberPriority ?? 0;
		this.subscriberOrdered = init.subscriberOrdered ?? 0;
		this.subscriberMaxLatency = init.subscriberMaxLatency ?? 0;
		this.startGroup = init.startGroup ?? 0;
		this.endGroup = init.endGroup ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.varint(this.subscribeId);
		e.string(this.broadcastPath);
		e.string(this.trackName);
		e.uint8(this.subscriberPriority);
		e.uint8(this.subscriberOrdered);
		e.varint(this.subscriberMaxLatency);
		e.varint(this.startGroup);
		e.varint(this.endGroup);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		let err: Error | undefined;

		// Read message length
		let msgLen: number;
		[msgLen, , err] = await readVarint(r);
		if (err) return err;

		// Read message body into a buffer
		const buf = new Uint8Array(msgLen);
		[, err] = await readFull(r, buf);
		if (err) return err;

		// Parse fields from the buffer
		const d = new MessageDecoder(buf);

		this.subscribeId = d.varint();
		this.broadcastPath = d.string();
		this.trackName = d.string();
		this.subscriberPriority = d.uint8();
		this.subscriberOrdered = d.uint8();
		this.subscriberMaxLatency = d.varint();
		this.startGroup = d.varint();
		this.endGroup = d.varint();

		return undefined;
	}
}
